Media table migration to react component

MibettyMeanie opened the pull request:
React component to render media table in multimedia page

Potential Bugs:

1. Submission date sorting might not be working properly
2. Page might not load properly in Safari

coder13 commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,119 @@
+    return ans;
+}
+
+function exampleReducer(state, action) {
```
It would be good to have proper naming of variables in this situation.

MibettyMeanie replied:
Fixed

coder13 commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
It looks like 4 width indentation was used when the standard so far has been to use 2 width indentation.

MibettyMeanie replied:
FIxed

coder13 commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,119 @@
+import { competitionUrl } from '../../lib/requests/routes.js.erb';
+import moment from 'moment';
+
+function dateRangeBetween(fromDate, toDate) {
```
Would you want to create a file in `utils/` for formatting date strings and start putting code like this there?

MibettyMeanie replied:
I have moved this to libs/helper as its already created used

coder13 commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,119 @@
+            if (state.column === action.column) {
+                return {
+                    ...state,
+                    data: state.data.slice().reverse(),
```
Unless this will be manipulated further (adding / removing items from this list), this can very likely be switched to a derived prop and we just compute it's state in the `MediaTable` component wrapping it with `useMemo`

dunkOnIT commented:
Hi @MibettyMeanie! Thanks for the PR on this. The main change I think we need to make is to move to serving the Media data via an API, rather than as a Rails object (`@media`) in the View. 

I've done my best to explain this [here](https://docs.worldcubeassociation.org/guides/passing_data_react_rails.html) - please let me know if any of that doesn't make sense, I'd be happy to help further!

Specifically, I think we'd need to make the following changes:
- [ ] Have `get_media`return a JSON of the data it fetches
- [ ] Create a /media API endpoint (it doesn't look like we have one at the moment) which returns the JSON that `get_media` creates
  - [ ] This endpoint should accept the parameters that `get_media` expects (`year`, `region`, `status`). I'd suggest that we pass these parameters as URL parameters
- [ ] Update your React code to fetch this data via the API call, rather than expecting to receive it immediately from the view (once it has the data, no change should be needed - the JSON format from the API can match pretty closely the one you get from the view at the moment)

Let me know if any of that doesn't make sense, or if there's anything I can do to help! I am _literally_ here to help :P

MibettyMeanie commented:
Thanks a lot for the review @coder13 and @dunkOnIT. I'll work on the changes you both suggested and will push it soon.

coder13 commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,171 @@
+
+export default function MediaTable({ isValidate }) {
+  const handleUpdateSuccess = () => {
+    setModalOpen(true);
```
Does this line of code work? Trying to access a variable that doesn't get defined until line 28?

MibettyMeanie commented:
> Hi @MibettyMeanie! Thanks for the PR on this. The main change I think we need to make is to move to serving the Media data via an API, rather than as a Rails object (`@media`) in the View.
> 
> I've done my best to explain this [here](https://docs.worldcubeassociation.org/guides/passing_data_react_rails.html) - please let me know if any of that doesn't make sense, I'd be happy to help further!
> 
> Specifically, I think we'd need to make the following changes:
> 
> * [ ]  Have `get_media`return a JSON of the data it fetches
> * [ ]  Create a /media API endpoint (it doesn't look like we have one at the moment) which returns the JSON that `get_media` creates
>   
>   * [ ]  This endpoint should accept the parameters that `get_media` expects (`year`, `region`, `status`). I'd suggest that we pass these parameters as URL parameters
> * [ ]  Update your React code to fetch this data via the API call, rather than expecting to receive it immediately from the view (once it has the data, no change should be needed - the JSON format from the API can match pretty closely the one you get from the view at the moment)
> 
> Let me know if any of that doesn't make sense, or if there's anything I can do to help! I am _literally_ here to help :P

Hey @dunkOnIT, I have added API referring to other controllers and it is now only for fetching and updating, the old controller has ruby API's which are used in the edit view (I haven't changed it in this PR). Also @danieljames-dj informed me that WCT leader has asked him not to delete rejected media from table rather than keep it intact with the status as rejected, so I have implemented like that.

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+  Dropdown,
+} from 'semantic-ui-react'
+export default function MediaTable({ isValidate }) {
+  const countryOptions = _.map(countries.byIso2, (country) => ({
+    key: country.iso2,
+    text: country.name,
+    value: country.iso2,
+  }));
+  const continentOptions = continents.map((continent) => ({
+    key: continent.id,
+    text: continent.name,
+    value: continent.name,
+  }));
+  const yearOptions = [
+    {
+      key: 'all',
+      text: 'All Years',
+      value: 'All Years',
+    },
+    ...years.map((year) => ({
+      key: year,
+      text: year,
+      value: year,
+    })),
+  ];
```
These could be defined outside the component (otherwise they're recomputed each re-render).

MibettyMeanie replied:
Fixed

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+  DropdownHeader,
+  DropdownDivider,
+  Dropdown,
+} from 'semantic-ui-react'
+export default function MediaTable({ isValidate }) {
```
```suggestion
} from 'semantic-ui-react'

export default function MediaTable({ isValidate }) {
```

MibettyMeanie replied:
Fixed

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+    setSelectedYear(data.value);
+  };
+  const handleSelectRegion = (event, data) => {
+    console.log(data)
```
Can this be removed?

MibettyMeanie replied:
Done

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+  );
+  const [modalOpen, setModalOpen] = useState(false);
+
+  useEffect(() => {
+    if (data) {
+      setMediaCombined(data?.map((medium, idx) => {
```
The current implementation isn't necessarily bad, but generally avoiding side effects and having fewer states is good. I'd suggest making `mediaCombined` state-less:
```
const mediaCombined = data?.map((medium, idx) => { ... })
```
and then having a separate (state-less, memoized) const for the sorted data:
```
const sortedMediaCombined = useMemo(
  () => [...mediaCombined]?.sort(...), // probably doesn't work exactly as written in the null case
  [mediaCombined, sortColumn, sortDirection],
]
```
and then using
```
{sortedMediaCombined?.map((media_row) => (
```
in the jsx return.

Thoughts?

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+      }))
+    }
+  }, [data]);
+  const [modalParams, setModalParams] = React.useState({
```
```suggestion
  const [modalParams, setModalParams] = useState({
```

MibettyMeanie replied:
Done

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+      sync(); handleUpdateSuccess(); setModalParams({ ...modalParams, open: false })
+    }, { method: 'PATCH' });
+  };
+  const [sortColumn, setSortColumn] = useState(null);
+  const [sortDirection, setSortDirection] = useState(null);
+  const handleSort = (clickedColumn) => () => {
+    if (sortColumn !== clickedColumn) {
+      setSortColumn(clickedColumn);
+      setSortDirection('ascending');
+      setMediaCombined([...mediaCombined].sort((a, b) => a[clickedColumn] - b[clickedColumn]));
+    } else {
+      setMediaCombined([...mediaCombined].reverse());
+      setSortDirection(sortDirection === 'ascending' ? 'descending' : 'ascending');
+    }
+  };
```
This could be encapsulated via `useReducer`:
```
const [{ sortColumn, sortDirection }, handleSort] = useReducer(sortReducer, { sortColumn: null, sortDirection: null })
```
(with a definition of `sortReducer` outside the component).

MibettyMeanie replied:
Done

kr-matthews commented on WcaOnRails/app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,257 @@
+          {mediaCombined?.map((media_row) => (
+            <Table.Row>
+              <Table.Cell>
+                {moment.utc(media_row.timestampSubmitted).format('MMMM DD, YYYY HH:mm UTC')}
```
I think we prefer to use luxon over moment? Someone correct me If I'm wrong.

MibettyMeanie replied:
Done

coder13 commented on app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,254 @@
+    value: year,
+  })),
+];
+handleSort = (clickedColumn) => () => {
```
This variable should be declared with `const`

kr-matthews replied:
This has the same name as the dispatch in the `userReducer` below. Also, you can't access these setters outside of the component (plus some of them don't exist). I don't think the reducer has been set up properly (but the intention seems right). I'm happy to help on slack if you want.

coder13 commented on app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,254 @@
+  DropdownDivider,
+  Dropdown,
+} from 'semantic-ui-react'
+const countryOptions = _.map(countries.byIso2, (country) => ({
```
super nit: would be good to see some newlines in between these variable declarations.

coder13 commented on app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,254 @@
+  const [modalOpen, setModalOpen] = useState(false);
+
+  useEffect(() => {
+    if (data) {
```
nit: I'd use a `useMemo` to transform the data rather than a useEffect unless you also need to modify this data in `mediaCombined` after initial fetching.

kr-matthews replied:
That is along the lines of what I suggested [here](https://github.com/thewca/worldcubeassociation.org/pull/8165#discussion_r1495037652); I didn't include a `useMemo` wrapper, but including one would be good.

kr-matthews commented on app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,254 @@
+      sync(); handleUpdateSuccess(); setModalParams({ ...modalParams, open: false })
+    }, { method: 'PATCH' });
+  };
+  const [{ sortColumn, sortDirection }, handleSort] = useReducer(sortReducer, { sortColumn: null, sortDirection: null })
```
Is `sortReducer` defined anywhere? I can't find it,

kr-matthews commented on app/webpacker/components/Media/MediaTable.jsx:
```diff
@@ -0,0 +1,254 @@
+          startDate: medium.competition.start_date,
+          endDate: medium.competition.end_date,
+          mediaId: medium.id,
+          competetionDate: dateRangeBetween(medium.competition.start_date, medium.competition.end_date),
```
"competetion" should be "competition"; six occurrences overall.
